File: tests/quaternion.py
from geovox.util import Quaternion
from geovox.linalg import Vector
from numpy import sin, cos
from numpy import pi as PI


U = Quaternion(1, Vector(2,3,4))
Q = Quaternion(2, Vector(-2,3,-4))

print(f"U= {repr(U)}")
print(f"Q= {repr(Q)}")

print("\nADDITION:")
print(f"U+Q= {repr(U+Q)}")
print(f"Q+U= {repr(Q+U)}")

print("\nSUBTRACTION:")
print(f"U-Q= {repr(U-Q)}")
print(f"Q-U= {repr(Q-U)}")

print("\nMULTIPLICATION")
print(f"U*Q= {repr(U*Q)}")
print(f"Q*U= {repr(Q*U)}")

print("\nSCALAR MULTIPLICATION:")
print(f"-2*U= {repr(-2*U)}")

print("\nINVERSE:")
print(f"U.inv()= {repr(U.inv())}")
print(f"U*U.inv()= {repr(U*U.inv())}")
print(f"U*Q.inv()= {repr(U*Q.inv())}")
print(f"U/Q = {repr(U/Q)}")

print("\nNEGATION:")
print(f"-U= {repr(-U)}")

print("\nSIZE:")
print(f"U.abs2= {U.abs2}")
print(f"abs(U)= {abs(U)}")
U.normalize()
print(f"U.normalize()-> U= {repr(U)}")
print(f"U.abs2= {U.abs2}")

print("\nASSIGNMENT:")
q0 = 3
qv = Vector(1,2,3)

print(f"Q= {repr(Q)}")
Q.q0 = q0
print(f"Q.q0= {q0} -> Q= {repr(Q)}")
Q.qv = qv
print(f"Q.qv= {qv} -> Q= {repr(Q)}")


print("\nROTATION")
theta = PI/4
axis  = Vector(0,0,3)
point = Vector(1,0,1)
Q.setrotation(theta, axis)
print(f"Q.setrotation({theta}, {repr(axis)})= {repr(Q)}")
print(f"Q.angle= {Q.angle}= PI/{PI/Q.angle}")
print(f"Q.axis= {repr(Q.axis)}")
print(f"point= {repr(point)}")
print(f"Q.rotate(point)= {repr(Q.rotate(point))}")

File: c_geovox/src/util/quaternion.cpp
#include "util/quaternion.hpp"

namespace GeoVox::util{
	double Quaternion::operator[](int idx) const{
		if (idx==0){
			return _q0;
		}
		return _qv[idx-1];
	}

	double Quaternion::q0() const {return _q0;}

	Point Quaternion::qv() const {return _qv;}

	Quaternion Quaternion::conj() const{
		return Quaternion(_q0, -_qv);
	}
	Quaternion Quaternion::inv() const{
		double C = 1.0/norm2();
		return Quaternion(C*_q0, (-C)*_qv);
	}
	double Quaternion::norm2() const{
		return _q0*_q0 + _qv.norm2();
	}
	double Quaternion::norm() const{
		return std::sqrt(norm2());
	}

	Quaternion* Quaternion::normalize(){
		double C = 1.0/norm();
		_q0*=C;
		_qv*=C;
		return this;
	}
	Quaternion* Quaternion::setrotation(const double& theta, const Point& axis){
		_q0 = std::cos(0.5*theta);
		_qv = std::sin(0.5*theta)*axis.normalize();
		return this;
	}
	Point Quaternion::rotate(const Point& point) const {
		Quaternion V = Quaternion(0.0, point);
		V = operator*((V*conj()));
		return V.qv();
	}

	///// ARITHMETIC
	Quaternion* Quaternion::operator+=(const Quaternion& other){
		_q0+=other.q0();
		_qv+=other.qv();
		return this;
	}
	Quaternion Quaternion::operator+(const Quaternion& other) const{
		return Quaternion(_q0+other.q0(), _qv+other.qv());
	}
	Quaternion* Quaternion::operator-=(const Quaternion& other){
		_q0-=other.q0();
		_qv-=other.qv();
		return this;
	}
	Quaternion Quaternion::operator-(const Quaternion& other) const{
		return Quaternion(_q0-other.q0(), _qv-other.qv());
	}
	Quaternion* Quaternion::operator*=(const Quaternion& other){
		double Q0 = _q0*other.q0() - _qv.dot(other.qv());
		_qv = _q0*other.qv() + other.q0()*_qv + _qv.cross(other.qv());
		_q0 = Q0;
		return this;
	}
	Quaternion Quaternion::operator*(const Quaternion& other) const{
		double Q0 = _q0*other.q0() - _qv.dot(other.qv());
		Point  QV = _q0*other.qv() + other.q0()*_qv + _qv.cross(other.qv());
		return Quaternion(Q0, QV);
	}
	Quaternion* Quaternion::operator/=(const Quaternion& other){
		operator*=(other.inv());
		return this;
	}
	Quaternion Quaternion::operator/(const Quaternion& other) const{
		return operator*(other.inv());
	}


}